### PR TITLE
Updater - Major refactor/cleanup of command classes. Refactor update actions to use new commands.

### PR DESCRIPTION
**Pull request by CoderBrandon**

### CODE REVIEW

File: src/Model/Cli/BaseCliCommand.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    public function __construct(array $args)
+    {
+        if (empty($args)) {
+            throw new Exception('Invalid Arguments: You must pass at least one argument.');
+        }
+        
+        $this->config = $this->getConfig();
+        
+        // @todo: check for config prefix
+        if (!empty($this->config->getCommandPrefix())) {
```
Comment (daceej) [inline src/Model/Cli/BaseCliCommand.php:47]: I like where this is going, but this is going to lock us into using a global prefix for all cli commands. That makes sense for using something like lando, but would prevent us from adding environment variables or anything else that may be specific to a particular tool we are using.

I have not thought through all of the scenarios that we may need to cover, so this is fine for now. I suspect this will need an update soon though.

File: src/Model/Cli/Runner/Drush.php [hunk 2]
```diff
@@ -249,9 +380,12 @@ private function ensureLocalSettings()
         $finder = new Finder();
         $finder->files();
         $finder->in(getcwd());
-        $finder->name('settings.local.php');
+        $filename = $this->config->getLocalSettingsFilename();
```
Comment (daceej) [inline src/Model/Cli/Runner/Drush.php:383]: What is the motivation behind this change? Is this to distinguish between sites that are older that may have a `secret.settings.php` file?

File: src/Model/Cli/Runner/Git.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    
+    /**
+     * Commit staged changes.
+     *
+     * @param string $message
+     *
+     * @return Process
+     * @throws Exception
+     */
+    public function commit($message = 'Committing changes.'): Process
```
Comment (daceej) [inline src/Model/Cli/Runner/Git.php:65]: Nitpick, but can you make `$message` required? I'd rather an exception be thrown than a commit message that does not accurately describe the change.

File: src/Model/Config/ProjectConfig.php [hunk 4]
```diff
@@ -44,9 +43,9 @@ class ProjectConfig
     public const KEY_TASKS = 'tasks';
     public const KEY_UPSTREAMS = 'upstreams';
     public const KEY_COMPOSER_PATH = 'composer_path';
-    public const KEY_DRUSH_MAJOR_VERSION = 'drush_major_version';
-    public const KEY_SYMFONY_CLI = 'symfony_cli';
-    public const KEY_DRUSH_PATCHER_INSTALLED = 'drush_patcher_installed';
+    public const KEY_COMMAND_PREFIX = 'command_prefix';
+    public const KEY_LOCAL_SETTINGS_FILENAME = 'local_settings_filename';
```
Comment (daceej) [inline src/Model/Config/ProjectConfig.php:47]: Same question from above about the local settings file.

File: src/Model/Cli/BaseCliCommand.php [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            array_unshift($args, $this->config->getCommandPrefix());
+        }
+        
+        $this->process = new Process($args);
+        $this->process->setTimeout($this->config->getTimeout());
```
Comment (daceej) [inline src/Model/Cli/BaseCliCommand.php:52]: Similar question here. How do you see this getting used?

File: src/Model/IO/IOStyle.php [hunk 6]
```diff
@@ -4,13 +4,146 @@
...
     // See https://github.com/symfony/console/blob/5.x/Style/SymfonyStyle.php.
+    
+    /**
+     *
+     * @param $message
+     * @param $command
+     *
+     * @return string
+     */
+    public function message($message, $command)
```
Comment (daceej) [inline src/Model/IO/IOStyle.php:21]: This is my bad -- when I stopped working on the refactor, this is where I left off since I knew it was going to break the updated code.

I would like to keep this file specifically for output so that we can define our own style of sorts. This would also be the place to handle any custom output that we want for something that is not supported out of the box (ie, tables, progress bars, etc...).

Instead of putting the stuff from `Runner` here, can you move them to a `CliHelper` class instead? I think that would be more intuitive in long term for these cases where we have a `Process` that we need to setup or do some funny business to get the output that we need. From there, then we can use the `IO` class to display the data to the user.

File: src/Model/Output/Runner.php.delete [hunk 7]
```diff
@@ -2,7 +2,6 @@
 
 namespace Waffle\Model\Output;
```
Comment (daceej) [inline src/Model/Output/Runner.php.delete:3]: I think this was committed by mistake (the move / rename).

**Review by daceej — CHANGES_REQUESTED** [review]
Overall it looks good, but there are a few things I think we should change.

Some of the items are good enough for now and we may just leave them until we run into a situation where it breaks.

If you don't mind, I'm happy to step in and make some of these changes (or provide clarification of what I'm thinking), just let me know.

File: src/Model/Cli/BaseCliCommand.php (hunk 1 above)
Comment (CoderBrandon) [reply src/Model/Cli/BaseCliCommand.php:47]: yeah, this was specifically for `lando` support, but I don't think it even works since I don't believe you need to put `lando` in front of Composer commands. It needs more testing/tweaking.

File: src/Model/Cli/Runner/Drush.php (hunk 2 above)
Comment (CoderBrandon) [reply src/Model/Cli/Runner/Drush.php:383]: I had a project where the local settings file was named something differently. This lets you define the secret/local settings file's filename per project.

File: src/Model/Config/ProjectConfig.php (hunk 4 above)
Comment (CoderBrandon) [reply src/Model/Config/ProjectConfig.php:47]: I had a project where the local settings file was named something differently. This lets you define the secret/local settings file's filename per project.

File: src/Model/Cli/BaseCliCommand.php (hunk 5 above)
Comment (CoderBrandon) [reply src/Model/Cli/BaseCliCommand.php:52]: I had a timeout parameter in the update tool but I was needing to pass it explicitly every time to each process I created. This instead lets you define a timeout globally instead for projects that need it. Some commands like `composer update` can take longer than the default timeout amount.

File: src/Model/IO/IOStyle.php (hunk 6 above)
Comment (CoderBrandon) [reply src/Model/IO/IOStyle.php:21]: Sure, that works. Where should `CliHelper` live and how should it be instantiated in the update tool? Should we put it in `BaseCommand`?

File: src/Model/Cli/BaseCliCommand.php (hunk 5 above)
Comment (daceej) [reply src/Model/Cli/BaseCliCommand.php:52]: Got it. Looks like you put in a default already, so this looks great.

File: src/Model/IO/IOStyle.php (hunk 6 above)
Comment (daceej) [reply src/Model/IO/IOStyle.php:21]: I think it would fit in fine either at `Waffle\Helper\CliHelper` or `Waffle\Model\Cli\CliHelper`, I don't really have any preference there.
